heron/schedulers/src/java/com/twitter/heron/scheduler/mesos/MesosLauncher.java

```diff
@@ -0,0 +1,147 @@
+    this.schedulerWorkingDirectory = MesosContext.getSchedulerWorkingDirectory(config);
+  }
+
+  @Override
+  public void close() {
+
+  }
+
+  @Override
+  public boolean launch(PackingPlan packing) {
+    // setup the scheduler working directory
+    // mainly it downloads and extracts the topology package
+    if (!setupWorkingDirectory()) {
+      LOG.severe("Failed to setup working directory");
+      return false;
+    }
+
+    String[] schedulerCmd = getSchedulerCommand();
+
+    Process p = startScheduler(schedulerCmd);
+
+    if (p == null) {
+      LOG.severe("Failed to start SchedulerMain using: " + Arrays.toString(schedulerCmd));
+      return false;
+    }
+
+    LOG.info(String.format(
+        "For checking the status and logs of the topology, use the working directory %s",
+        MesosContext.getSchedulerWorkingDirectory(config)));
+
+    return true;
+  }
+
+  protected String[] getSchedulerCommand() {
+    List<Integer> freePorts = new ArrayList<>(SchedulerUtils.PORTS_REQUIRED_FOR_SCHEDULER);
+    for (int i = 0; i < SchedulerUtils.PORTS_REQUIRED_FOR_SCHEDULER; i++) {
+      freePorts.add(SysUtils.getFreePort());
+    }
+
+    List<String> commands = new ArrayList<>();
+
+    // The java executable should be "{JAVA_HOME}/bin/java"
+    String javaExecutable = String.format("%s/%s", Context.javaHome(config), "bin/java");
+    commands.add(javaExecutable);
+
+    // Add the path to load mesos library
+    commands.add(String.format("-Djava.library.path=%s",
+        MesosContext.getHeronMesosNativeLibraryPath(config)));
+    commands.add("-cp");
+
+    // Construct the complete classpath to start scheduler
+    String completeSchedulerProcessClassPath = System.getProperty("java.class.path");
+    commands.add(completeSchedulerProcessClassPath);
+    commands.add("com.twitter.heron.scheduler.SchedulerMain");
+
+    // Add default scheduler properties override
+    // Add the topology package uri as a String and pass the SchedulerMain as a property
+    commands.add(String.format("-%s%s=%s",
+        Keys.SCHEDULER_COMMAND_LINE_PROPERTIES_OVERRIDE_OPTION,
+        Keys.topologyPackageUri(),
+        Runtime.topologyPackageUri(runtime).toString()));
+
+    String[] commandArgs = schedulerCommandArgs(freePorts);
+    commands.addAll(Arrays.asList(commandArgs));
+
+    return commands.toArray(new String[0]);
+  }
+
+  ///////////////////////////////////////////////////////////////////////////////
+  // Utils methods leveraging unit tests
+  ///////////////////////////////////////////////////////////////////////////////
+  protected boolean setupWorkingDirectory() {
+    if (!FileUtils.isDirectoryExists(schedulerWorkingDirectory)) {
+      FileUtils.createDirectory(schedulerWorkingDirectory);
+    }
+
+    String topologyPackageURL = String.format("file://%s", Context.topologyPackageFile(config));
+    String topologyPackageDestination = Paths.get(
+        schedulerWorkingDirectory, "topology.tar.gz").toString();
```

Comment:
Can the topology package name be picked up from config? For e.g. `Context.topologyPackageFile(config)`